### PR TITLE
Refactored tests to use pytest if require <= 1 GPU

### PR DESCRIPTION
**Pull request by MChoi-git**
PR which creates delineation between cpu/single-gpu (`pytest`) tests and multi-gpu (internally-run `submitit`) tests. ~~Single-gpu/cpu tests are included in this PR. The `submitit` tests will be supported in a future PR.~~

Support for multi-gpu tests run internally via `submitit` added. Some multi-gpu tests are still pending, since they require huggingface accelerate plugins/tooling which isn't supported by `submitit`. I am thinking that we refactor the huggingface FSDP tests to simply use native `torch.distributed.FSDP` instead.

### CODE REVIEW

File: tests/core/test_hook_function.py [hunk 1]
```diff
@@ -1,16 +1,30 @@
...
 # to other kinds of layers
-MODULE_NAME = "model.layers.27.mlp"
+MODULE_NAME = "model.decoder.layers.9.fc2"
+
+
+def rgetattr(module, attr):
+    def _getattr(module, attr):
+        return getattr(module, attr)
+
+    return reduce(_getattr, [module] + attr.split("."))
```
Comment (adil-a) [inline tests/core/test_hook_function.py:22]: Why can't we just call `getattr` directly? What's the use of a private function here?

File: tests/core/test_hook_function.py [hunk 2]
```diff
@@ -38,11 +52,9 @@ def __call__(self, *args, **kwargs) -> torch.Tensor:
         return out
 
 
-def inject_module(injection_layer: nn.Module, model: nn.Module) -> None:
+def inject_module(injection_layer: nn.Module, model: nn.Module) -> nn.Module:
     """Replace <MODULE_NAME> in <model> with <injection_layer>"""
-    sliced = MODULE_NAME.split(".")
-    layer_num = int(sliced[2])
-    model.model.layers[layer_num].mlp = injection_layer
+    rsetattr(model, MODULE_NAME, injection_layer)
```
Comment (adil-a) [inline tests/core/test_hook_function.py:57]: This function returns `None` as per the documentation for `setattr`

File: tests/core/test_wrapper.py [hunk 3]
```diff
@@ -16,37 +19,12 @@
...
     """
     Tests if a hook function is registered correctly, and if the fields are set
     appropriately.
     """
-    model = make_model()
-    model.cuda()
+    model = opt_350m.cuda()
 
     activations = {}
     model = FlexModel(model, activations,)
```
Comment (adil-a) [inline tests/core/test_wrapper.py:30]: Remove `,` after after the `activations` argument

File: tests/core/test_wrapper.py [hunk 4]
```diff
@@ -174,8 +148,51 @@ def test_destroy():
...
+    model = FlexModel(model, activations)
+
+    prompts = [
+        "It's a nice day we're having",
+        "The capital of Canada is",
+        "What should I eat for dinner tonight?",
+        "There's about three people going to",
+    ]
+
+    inputs = tokenizer(prompts, padding=True, return_tensors="pt",)["input_ids"].cuda()
```
Comment (adil-a) [inline tests/core/test_wrapper.py:166]: Remove `,` after the `return_tensors` argument

**Review by adil-a — APPROVED** [review]
Slight changes, LGTM otherwise

File: tests/core/test_hook_function.py (hunk 1 above)
Comment (MChoi-git) [reply tests/core/test_hook_function.py:22]: `getattr` and `setattr` unfortunately don't work recursively, ie. to get the module name.

**Pull request merged**